### PR TITLE
Add Flask Support

### PR DESCRIPTION
**Pull request by johnchuks**
## Description

This PR adds a `Flask` view that exposes a `strawberry` schema.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Added a Flask view that allows you to query the schema and interact with it via GraphQL playground.

Usage:

```python
# app.py
from strawberry.flask.views import GraphQLView
from your_project.schema import schema

app = Flask(__name__)

app.add_url_rule(
    "/graphql", view_func=GraphQLView.as_view("graphql_view", schema=schema)
)

if __name__ == "__main__":
    app.run(debug=True)
```

File: pyproject.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ uvicorn = "=0.10.0"
 django = {version = "^2.0",optional = true}
 graphql-core = {version = "^3.0.0a0", allows-prereleases = true}
 aniso8601 = "^8.0.0"
+flask = "^1.1"
```
Comment (patrick91) [inline pyproject.toml]: can you add this to the extras? Like we've done for django? 😊

File: strawberry/flask/playground.py [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+      GraphQLPlayground.init(root, {
+        endpoint: "{{ request_path }}",
+        codeTheme: colors
+      })
+    })
+  </script>
+</body>
+</html>
+
+"""
```
Comment (patrick91) [inline strawberry/flask/playground.py]: we have this here already :) https://github.com/strawberry-graphql/strawberry/blob/master/strawberry/static/playground.html

File: strawberry/flask/tests/schema.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+
+@strawberry.type
+class Query:
+    @strawberry.field
+    def user(self, info) -> User:
+        return User(name="Patrick", age=100)
+
+
+schema = strawberry.Schema(query=Query)
```
Comment (patrick91) [inline strawberry/flask/tests/schema.py]: I think we have this somewhere else already, can we make it into a fixture (maybe it already is) and reuse it? :)

Comment (patrick91) [reply strawberry/flask/tests/schema.py]: should be in the tests for django I think

File: strawberry/flask/tests/test_graphqlview.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import pytest
+
+from flask import url_for
+
+from .app import init_app
+
+
+@pytest.fixture
+def app():
+    return init_app()
```
Comment (patrick91) [inline strawberry/flask/tests/test_graphqlview.py]: can we move the contents of init_app here? feels redundant to have another file for this ;)

we could move this fixture tuore flast/tests/conftest.py as well :)

File: strawberry/flask/tests/test_graphqlview.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def test_graphql_query(client):
+    query = {"query": "query {\n  user {\n    name\n    age\n  }\n}"}
```
Comment (patrick91) [inline strawberry/flask/tests/test_graphqlview.py]: let's make this into a multiline string, should be easier to read ;)

File: strawberry/flask/views.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    methods = ["GET", "POST"]
+
+    def __init__(self, schema=None):
+        self.schema = schema
+
+        if not self.schema:
+            raise ValueError("You must pass in a schema to GraphQLView")
+
+        if not isinstance(self.schema, GraphQLSchema):
+            raise ValueError("A valid schema is required to be provided to GraphQLView")
```
Comment (patrick91) [inline strawberry/flask/views.py]: 👍

File: strawberry/flask/views.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+class GraphQLView(View):
+    schema = None
+
+    methods = ["GET", "POST"]
+
+    def __init__(self, schema=None):
+        self.schema = schema
+
+        if not self.schema:
+            raise ValueError("You must pass in a schema to GraphQLView")
```
Comment (patrick91) [inline strawberry/flask/views.py]: I'm wondering if we should make this non-optional if we raise an error :)

Comment (johnchuks) [reply strawberry/flask/views.py]: Will do

File: strawberry/flask/tests/test_graphqlview.py (hunk 5 above)
Comment (johnchuks) [reply strawberry/flask/tests/test_graphqlview.py]: Will do

File: strawberry/flask/tests/test_graphqlview.py (hunk 4 above)
Comment (johnchuks) [reply strawberry/flask/tests/test_graphqlview.py]: Sure. I will move it to `init_app` method

File: pyproject.toml (hunk 1 above)
Comment (johnchuks) [reply pyproject.toml]: Sure will do

**Comment by jaydenwindle** [comment]
Looks great @johnchuks!

The Django integration (#137) currently lives in `strawberry/contrib/django/`. We should probably either move the Django integration to `strawberry/django` or move the flask integration to `strawberry/contrib/flask` for consistency.

@patrick91 any preference? I'm leaning towards `strawberry/django` and `strawberry/flask` to keep import strings short.

File: poetry.lock [hunk 8]
```diff
@@ -679,11 +827,16 @@ version = "0.6.0"
 [package.dependencies]
 more-itertools = "*"
 
+[package.extras]
+docs = ["sphinx", "jaraco.packaging (>=3.2)", "rst.linker (>=1.9)"]
+testing = ["pathlib2", "contextlib2", "unittest2"]
+
 [extras]
 django = ["django"]
+flask = ["flask"]
```
Comment (johnchuks) [inline poetry.lock]: Hey @jaydenwindle @patrick91 Poetry does not recognise optional dependencies in `dev-dependencies` when installing with `--extras`  flag as shown above. The only workaround is moving `pytest-django` and `pytest-flask` to `dependencies.

**Comment by johnchuks** [comment]
I agree with you @jaydenwindle I am also leaning more towards `strawberry/flask` and `strawberry/django`

File: poetry.lock (hunk 8 above)
Comment (patrick91) [reply poetry.lock]: @johnchuks I think it is fine to have all the dev dependencies always declared :) in dev we always want to test flask and django :)

File: RELEASE.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
+Release type: patch
```
Comment (patrick91) [inline RELEASE.md]: let's make this a minor release :)

File: strawberry/flask/tests/test_graphqlview.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import json
```
Comment (patrick91) [inline strawberry/flask/tests/test_graphqlview.py]: can you move this file under `/tests`? :)

File: strawberry/flask/tests/test_graphqlview.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    app.debug = True
+    app.add_url_rule(
+        "/graphql", view_func=GraphQLView.as_view("graphql_view", schema=schema)
+    )
+    return app
+
+
+if __name__ == "__main__":
+    test_app = init_app()
+    test_app.run()
```
Comment (patrick91) [inline strawberry/flask/tests/test_graphqlview.py]: we don't need this for tests, right? :)

File: strawberry/flask/tests/test_graphqlview.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+
+@pytest.fixture
+@pytest.mark.usefixtures("schema")
```
Comment (patrick91) [inline strawberry/flask/tests/test_graphqlview.py]: don't think we need this with functions: https://docs.pytest.org/en/latest/reference.html#pytest-mark-usefixtures

File: strawberry/flask/views.py (hunk 7 above)
Comment (patrick91) [reply strawberry/flask/views.py]: no, I think we can remove `=None` from the arguments :)

image

Comment (johnchuks) [reply strawberry/flask/views.py]: Alright sounds good. Will do

File: RELEASE.md (hunk 9 above)
Comment (johnchuks) [reply RELEASE.md]: Will do

File: strawberry/flask/tests/test_graphqlview.py (hunk 10 above)
Comment (johnchuks) [reply strawberry/flask/tests/test_graphqlview.py]: Sure

File: strawberry/flask/tests/test_graphqlview.py (hunk 12 above)
Comment (johnchuks) [reply strawberry/flask/tests/test_graphqlview.py]: @patrick91 please could you explain more on this. I added a fixture for a generic schema that came be used in our test environment. It is located at `tests/fixtures/schema.py`

File: strawberry/flask/tests/test_graphqlview.py (hunk 11 above)
Comment (johnchuks) [reply strawberry/flask/tests/test_graphqlview.py]: We need it to create a flask app instance so we can run the tests. As you can see in `line 25` we are calling the test app.

**Comment by patrick91** [comment]
@johnchuks just merged, sorry it took me a while to get to it :)

**Pull request merged**